### DISCUSSION TITLE
Old Nesting Combinations

### DISCUSSION THREAD

**Discussion by Mossbraker** [adieyal/sd-dynamic-prompts#375] [Q&A]
I'm not actually sure if the old nesting combinations worked like this because when I started using this extension the `{2$$__wildcard__}` functionality wasn't working temporarily due to the changes that were being implemented, but is there a way to do something like 
`{2$$(__person__:__number__)}`
and have it choose one value from each of person.txt and number.txt twice while guaranteeing no duplicates? E.g.:
```
(person1:number3), (person2:number4)
```
Or what about if I want to have it choose one value from each of person.txt and number.txt twice while guaranteeing no duplicates from person.txt are used but allowing duplicates from number.txt to be used? Potentially allowing, for example:
```
(person1:number3), (person2:number3)
```
Currently, the stated template, i.e.:
`{2$$(__person__:__number__)}`
only returns a single combination of values from person.txt and number.txt. E.g.:
```
(person1:number3)
```

Also, is there a way to do this:
`{4-5$$(__person__|__number__)}`
and have it choose four or five values, any of which could be from person.txt or number.txt? Currently it only returns two values.

**Comment by adieyal** [comment]
is there a way to do something like `{2$$(__person__:__number__)}` and have it choose one value from each of person.txt and number.txt twice while guaranteeing no duplicates? 

Unfortunately not - the duplication is currently only checked in the case of `{2$$__wildcard__}`

> Or what about if I want to have it choose one value from each of person.txt and number.txt twice while guaranteeing no duplicates from person.txt are used but allowing duplicates from number.txt to be used? Potentially allowing, for example:
> 
> ```
> (person1:number3), (person2:number3)
> ```

again no

> Currently, the stated template, i.e.: `{2$$(__person__:__number__)}` only returns a single combination of values from person.txt and number.txt. E.g.:

The way it is current interpreted is as follows. Suppose you did this `{2$$red|green|blue|yellow}` - then it will choose 2 colours from the list. If your prompt looks like this:  `{5$$red|green|blue|yellow}` - then it will choose all 4 colours in random order.

When your variant is `{2$$__wildcard1__ __wildcard2__}`. It gets parsed 
`sequence = __wildcard1__ __wildcard2__`

and then
`{2$$sequence}`

So it will first resolve sequence by getting a random value from each wildcard, then it will get the value from `sequence`. This last bit is effectively the same as `{2$$red}`. So you can't sample it twice.

> Also, is there a way to do this: `{4-5$$(__person__|__number__)}` and have it choose four or five values, any of which could be from person.txt or number.txt? Currently it only returns two values.

Same as above, `{4-5$$__person__|__number__}` is handled in the same way as `{4-5$$red|green}`.

Your use-cases are legitimate, but not currently supported given how the prompt is resolved. That may change in future, in which case I'll update you. You might want to consider jinja2 templates - I know it's not the ideal answer, but that's the only thing I can suggest to achieve what you want right now.

**Comment by Mossbraker** [comment]
Thanks for the response. I was playing around with jinja2 templates before and they started to get complicated rather quickly and I don't think my code was very efficient (as I have little experience writing scripts).

So, using jinja2 templates, how would I do something like either `{4-5$$__person__|__number__}` or `{2$$(__person__:__number__)}` and have it choose the stated number of options from person.txt and number.txt while guaranteeing no duplicates from either wildcard? What about just guaranteeing no duplicates from one of the wildcards?

What about if I want to do something like:
`{2-3$$ $$__flower__|0.5::__number__} outside in a garden and {2$$ and $$__number__ __flower__} in a vase` 
and guarantee no wildcard option is ever repeated for the entirety of the prompt (so, even though the wildcards are called at both the beginning and the end of the prompt, no flowers or numbers are ever repeated) using jinja2 templates?
